elytron/WFCORE-4058-automatic-keystore-detection.adoc

```diff
@@ -0,0 +1,92 @@
+= Automatic detection of file based KeyStore types.
+:author:            Maritn Mazanek
+:email:             mmazanek@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+* If a user does not specify the type of a file based KeyStore can we auto detect the type?
```

Comment:
Idea: What about making this opt-in feature. E.g. provide custom type value, e.g.  "AUTO" or "DETECT". I am still worrying about making this code branch default, so everybody must go through by default.

Reply:
Maybe this should be clarified in the analysis but this should definitely be opt-in - i.e. if no type specified we use discovery of the type.  If they specify a type that should be the only type we try so this will not impact any existing default configurations out there.

Reply:
This would be only for the keystores without defined type, as is required before this change. You would basically opt-in by not specifying the type. (thought is that a good idea? - maybe put keystore type as "auto" or something like that, with the hope that it wouldn't collide with actual keystore type name)

Reply:
by specifying the keystore type as some hardcoded string (let's say "auto") we would not need to make a model change, only issue would be the collision with probably not ever existing keystore type

Reply:
One of the motivations for an RFE like this is to reduce the configuration required - IMO if you don't define a value at all it should be expected we would have to work it out.  

If we start using strings that could collide with custom keystore implementations we will not have reduced the configuration required and now we will have introduced a new problem.

Reply:
You are looking from migration point of view. I am looking from starting from scratch point of view.

Once we change model to be key-store.type optional. This will be possible to create keystore
`/subsystem=elytron/key-store=a:add(credential-reference={clear-text=a},path=/path/to/keystore)`
So by default autodetection will be turned on. And to disable autodetection user action will be needed - type has to be specified. 

What I have been proposing is by default turn autodetection off and turn it on after user action - e.g. type="AUTO".

Reply:
The two proposals however are the same: -

1. don't define a type, we default to 'auto' and we detect the keystore type.
2. don't define a type, we take it as undefined so we detect the keystore type.

The only difference is we have now given the string 'auto' special meaning,

Reply:
No I meant add 'auto' to enum of option. Not to default to 'auto'
1. don't define a type, type is mandatory, it will fail
2. define auto, detect the keystore type

If detection become default, keep in mind startup time of server. Could detection be performed on first usage instead of startup?

Reply:
Yes I do not like the idea of using a type of 'auto' as having a special meaning as it would mean we could never support a custom KeyStore of type 'auto'.